_posts/frameworks/2016-12-19-spring-restdocs.md

```diff
@@ -0,0 +1,274 @@
+request is passed into the `document()` method to automatically create documentation
+for the request. The `document()` method is statically imported from the class
+`MockMvcRestDocumentation` to keep the code readable.
+
+The `MockMvc` object is initialized with a `JUnitRestDocumentation` object, as shown in the next code snippet. 
+This way, the `MockMvc` object is instrumented to create [Asciidoctor](http://asciidoctor.org/)
+snippets into the folder `build/generated-snippets`.
+
+```java
+private MockMvc mvc;
+
+@Rule
+public JUnitRestDocumentation restDocumentation = new JUnitRestDocumentation("build/generated-snippets");
+
+@Before
+public void setup() {
+    mvc = MockMvcBuilders.webAppContextSetup(applicationContext)
+            .apply(MockMvcRestDocumentation.documentationConfiguration(this.restDocumentation))
+            .build();
+}
+
+protected MockMvc mvc() {
+    return mvc;
+}
+
+```
+
+When the test is executed, Spring Rest Docs will now generate snippets into the snippets folder
+that contain an example request and an example response. The following snippets would be generated
+into the folder `build/generated-snippets/projects/create`.
+
+http-request.adoc:
+
+```
+[source,http,options="nowrap"]
+----
+POST /projects HTTP/1.1
+Content-Type: application/json
+Host: localhost:8080
+Content-Length: 129
+
+{
+  "name" : "name",
+  "vcsType" : "GIT",
+  "vcsUrl" : "http://valid.url",
+  "vcsUser" : "user",
+  "vcsPassword" : "pass"
+}
+----
+```
+
+http-response.adoc:
+
+```
+[source,http,options="nowrap"]
+----
+HTTP/1.1 201 Created
+Content-Type: application/hal+json;charset=UTF-8
+Content-Length: 485
+
+{
+  "name" : "name",
+  "vcsType" : "GIT",
+  "vcsUrl" : "http://valid.url",
+  "vcsUser" : "user",
+  "vcsPassword" : "pass",
+  "_links" : {
+    "self" : {
+      "href" : "http://localhost:8080/projects/1"
+    },
+    "files" : {
+      "href" : "http://localhost:8080/projects/1/files"
+    },
+    "analyzers" : {
+      "href" : "http://localhost:8080/projects/1/analyzers"
+    },
+    "strategy" : {
+      "href" : "http://localhost:8080/projects/1/strategy"
+    }
+  }
+}
+----
+```
+
+These examples already go a long way to documenting your REST API. Examples are the best way
+for developers to get to know your API. The snippets automatically generated from your test
+don't help when they rot in your snippets folder, though, so we have to expose them by 
+including them into a central documentation of some sorts.
+
+## Creating API Docs with Asciidoctor
+With the snippets at hand, we can now create our API documentation. The snippets are in 
+[Asciidoctor](http://asciidoctor.org/) format by default. Asciidoctor is a markup language 
+similiar to Markdown, but much more powerful. You can now simply create an Asciidoctor 
+document with your favorite text editor. That document will provide the stage for including the 
+snippets. An example document would look like this:
+
+```asciidoctor
```

Comment:
currently there is no support for asciidoctor in rouge: https://github.com/jneen/rouge/tree/master/lib/rouge/lexers